src/parser/mod.rs

```diff
@@ -45,3 +54,33 @@ impl<R: Read> Context<R> {
         &mut self.reader
     }
 }
+
+pub fn verify_starting_tag<R: Read>(
+    context: &mut Context<R>,
+    local_name: &'static str,
+) -> Result<(Vec<OwnedAttribute>)> {
+    //makes sure the specified starting tag is the next tag on the stream
+    //we ignore and skip all xmlevents except StartElement, Characters and EndElement
+    loop {
+        let next = context.reader.next();
+        match next {
+            Some(Ok(XmlEvent::StartElement {
+                name, attributes, ..
+            })) => {
+                ensure!(
+                    name.local_name == local_name,
+                    ErrorKind::InvalidChildElement(name.local_name, local_name)
```

Comment:
Technically speaking, this would give an error of "invalid child element 'wpt' in 'wpt'", right? Because we don't keep track of the parent element?

Or would this even happen because we check the name beforehand, so instead we should be dying here or something?

Reply:
I dont think it should be wpt in wpt, note the first is name.local_name, which is the actual starting tag we found, and the other is "local_name", which is the expected local_name (the starting tag we are looking for). So I think what would happen, when looking for the wpt starting tag and getting trk starting tag instead is "invalid child element trk in wpt".

I agree this is stretching the InvalidChildElement error a bit, we could add a InvalidStartingElement error that would read something like "invalid starting element trk found, expected wpt". Note that the way the parser currently functions it shouldnt be possible to produce this error except in gpx - because all other parsers are sub-parser, which would only be called if the parent parser already found their starting tag via peek().

Reply:
I see, that makes sense. It shouldn't need changing then, at least for now. Thanks for the clarity.